### PR TITLE
Dealing with typeError: must be str, not bytes

### PR DESCRIPTION
**Pull request by FMCumhaill**
Dealing with typeError: must be str, not bytes which occurred for me when running this file. https://stackoverflow.com/questions/5512811/builtins-typeerror-must-be-str-not-bytes

This may be a local requirement and therefore not sufficient for a pull request?

### CODE REVIEW

**Comment by piskvorky** [comment]
All I/O writing or reading should definitely be happening in binary mode (`rb`, `wb`). With explicit encoding conversions where necessary. This is true for both Python 2 and Python 3.

@loretoparisi was there a particular reason for the `w+`? Can we replace it with `wb`?

One more request: @FMCumhaill can you please replace all calls to `open()` with `smart_open()` (`from smart_open import smart_open`)?

The reason is that in Gensim, we're trying to abstract away from the storage, so people can save files to S3, HDFS, compressed etc, using the [smart_open library](https://github.com/RaRe-Technologies/smart_open).

**Comment by FMCumhaill** [comment]
@piskvorky done as requested! 👍

File: gensim/scripts/word2vec2tensor.py [hunk 1]
```diff
@@ -67,8 +68,8 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
     outfiletsv = tensor_filename + '_tensor.tsv'
     outfiletsvmeta = tensor_filename + '_metadata.tsv'
 
-    with open(outfiletsv, 'w') as file_vector:
-        with open(outfiletsvmeta, 'wb') as file_metadata:
+    with smart_open(outfiletsv, 'w') as file_vector:
```
Comment (piskvorky) [inline gensim/scripts/word2vec2tensor.py]: No `w` -- always use binary I/O (`rb`, `wb`). Encode text to utf8 where necessary.

Comment (FMCumhaill) [reply gensim/scripts/word2vec2tensor.py]: Changing that w to wb gives me the following error: 

Traceback (most recent call last):
  File "word2vec2tensor.py", line 94, in <module>
    word2vec2tensor(args.input, args.output, args.binary)
  File "word2vec2tensor.py", line 76, in word2vec2tensor
    file_vector.write(vector_row + '\n')
TypeError: a bytes-like object is required, not 'str'

Comment (piskvorky) [reply gensim/scripts/word2vec2tensor.py]: Yes -- please encode text to binary (utf8) where necessary.

File: gensim/scripts/word2vec2tensor.py [hunk 2]
```diff
@@ -44,6 +44,7 @@
 import argparse
 
 import gensim
+from smart_open import smart_open
```
Comment (menshikh-iv) [inline gensim/scripts/word2vec2tensor.py]: please add a test for `word2vec2tensor` function (for avoid this issue in future)

File: gensim/scripts/word2vec2tensor.py [hunk 3]
```diff
@@ -67,8 +68,8 @@ def word2vec2tensor(word2vec_model_path, tensor_filename, binary=False):
     outfiletsv = tensor_filename + '_tensor.tsv'
     outfiletsvmeta = tensor_filename + '_metadata.tsv'
 
-    with open(outfiletsv, 'w+') as file_vector:
-        with open(outfiletsvmeta, 'w+') as file_metadata:
+    with smart_open(outfiletsv, 'w') as file_vector:
```
Comment (menshikh-iv) [inline gensim/scripts/word2vec2tensor.py]: must be `wb`

Comment (menshikh-iv) [reply gensim/scripts/word2vec2tensor.py]: also, good idea to make a  code more "flat", like this
```python
with smart_open(...) as file_vector, smart_open(...) as file_metadata:
    ...
```

**Comment by menshikh-iv** [comment]
@FMCumhaill thanks for PR, do you plan to finish it?

File: gensim/scripts/word2vec2tensor.py (hunk 3 above)
Comment (FMCumhaill) [reply gensim/scripts/word2vec2tensor.py]: @menshikh-iv  sorry this escaped me! ATM I don't have time to finish it sorry. I have no idea how to add a test and don't have time to research it. My apologies!

Comment (menshikh-iv) [reply gensim/scripts/word2vec2tensor.py]: @FMCumhaill That's sad, but anyway, thanks for the answer, good luck!

**Comment by menshikh-iv** [comment]
Close by https://github.com/RaRe-Technologies/gensim/pull/2125#discussion_r206426486 reason

**Comment by piskvorky** [comment]
@menshikh-iv this is a bug, please don't close like this. We need to fix this either way (and the fix looks really trivial).

@loretoparisi was there a particular reason for the `w+`? Can we replace it with `wb`?

**Comment by menshikh-iv** [comment]
@piskvorky we need fix bugs, I agree, but this is PR, not a issue, if contributor don't have time to complete (and isn't close to finish) - I close it without "almost complete" badge.

Also, we already have opened issue for this bug #1958 

About "amost_complete" - I don't agree, because:
- incorrect file modes
- missing tests
- 3 lines diff where correct only one (import), I don't think that we should mark it with badge

**Comment by piskvorky** [comment]
OK. Let me remove the label.

**Pull request closed without merging**